src/core/service.c

```diff
@@ -188,7 +188,7 @@ void service_close_socket_fd(Service *s) {
         s->socket_fd = asynchronous_close(s->socket_fd);
 
         if (UNIT_ISSET(s->accept_socket)) {
-                socket_connection_unref(SOCKET(UNIT_DEREF(s->accept_socket)));
+                socket_connection_unref(SOCKET(UNIT_DEREF(s->accept_socket)), s);
                 unit_ref_unset(&s->accept_socket);
         }
```

Comment:
when the service is freed it should really release its "peer" refernce should it still have one. i.e. there should be socket_peer_unref() and it should be called from service_done() on the service's peer object.